ocw/lib/gke.py

```diff
@@ -0,0 +1,77 @@
+import os
+import subprocess
+from datetime import datetime, timezone
+import kubernetes
+import googleapiclient.discovery
+from google.oauth2 import service_account
+from ocw.lib.gce import GCE
+
+
+class GKE(GCE):
+    __instances = {}
+
+    def __new__(cls, vault_namespace):
+        if vault_namespace not in GKE.__instances:
+            GKE.__instances[vault_namespace] = self = object.__new__(cls)
+            self.__gke_client = None
+            self.__kubectl_client = {}
+            self.init()
+        return GKE.__instances[vault_namespace]
+
+    def get_project_id(self):
+        return self.get_data()["project_id"]
+
+    def gke_client(self):
+        if self.__gke_client is None:
+            self.private_key_data = self.get_data()
+            credentials = service_account.Credentials.from_service_account_info(self.private_key_data)
+            self.__gke_client = googleapiclient.discovery.build('container', 'v1', credentials=credentials)
+
+        return self.__gke_client
+
+    def kubectl_client(self, zone, cluster):
+        cluster_name = cluster["name"]
+        kube_dir = "/root/.kube"
+        kubeconfig = f"{kube_dir}/gke_config_{zone}_{cluster_name}"
+        try:
+            self.cmd_exec(f"gcloud auth login  --project={self.get_project_id()} --cred-file=/var/pcw/qac/GKE.json --quiet")
+            self.cmd_exec(f"gcloud container clusters get-credentials {cluster_name} " +
+                          f"--zone {zone} --project {self.get_project_id()}", {"KUBECONFIG": kubeconfig})
+            if not os.path.exists(kubeconfig):
+                raise FileNotFoundError("{kubeconfig} doesn't exists")
+        except subprocess.CalledProcessError as exception:
+            self.log_err(f"Cannot get the kubeconfig for the cluster {cluster_name} on region {zone}, reason: {str(exception)}")
+            return None
+
+        kubernetes.config.load_kube_config(config_file=kubeconfig)
+        self.__kubectl_client[zone] = kubernetes.client.BatchV1Api()
+        return self.__kubectl_client[zone]
+
+    def get_clusters(self, zone):
+        request = self.gke_client().projects().zones().clusters().list(projectId=self.get_project_id(), zone=zone)
+        response = request.execute()
+        if 'clusters' in response:
+            return response["clusters"]
+
+        return []
+
+    def cleanup_k8s_jobs(self):
+        for region in self.list_regions():
+            for zone in self.list_zones(region):
+                for cluster in self.get_clusters(zone):
+                    cluster_name = cluster["name"]
+                    self.log_info(f"Clean up of cluster {cluster_name} in zone {zone}")
+                    client = self.kubectl_client(zone, cluster)
+                    if client is not None:
```

Comment:
so if there is issue with client creation we would keep silent ?  please remove this `if` and just call `self.kubectl_client(zone, cluster).list_job_for_all_namespaces(watch=False)` I want to see exception if we can not do cleanup

Reply:
yes, it is really true

Reply:
but I remove the None I have a lint error
`ocw/lib/gke.py:29:4: R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)`

Maybe the best solution could be remove the try and allowing that when an error raises without caching